fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/IndexScrubbing.java

```diff
@@ -0,0 +1,247 @@
+                        .build();
+    }
+
+    @Override
+    CompletableFuture<Void> buildIndexInternalAsync() {
+        return getRunner().runAsync(context ->
+                        context.getReadVersionAsync()
+                                .thenCompose(vignore -> {
+                                    SubspaceProvider subspaceProvider = common.getRecordStoreBuilder().getSubspaceProvider();
+                                    return subspaceProvider.getSubspaceAsync(context)
+                                            .thenCompose(subspace ->
+                                                    indexScrub(subspaceProvider, subspace));
+                                }),
+                common.indexLogMessageKeyValues("IndexScrubbing::buildIndexInternalAsync"));
+    }
+
+    @Nonnull
+    private CompletableFuture<Void> indexScrub(@Nonnull SubspaceProvider subspaceProvider, @Nonnull Subspace subspace) {
+
+        final List<Object> additionalLogMessageKeyValues = Arrays.asList(LogMessageKeys.CALLING_METHOD, "indexScrub");
+
+        return iterateAllRanges(additionalLogMessageKeyValues,
+                this::indexScrubRangeOnly,
+                subspaceProvider, subspace);
+    }
+
+    @Nonnull
+    private CompletableFuture<Boolean> indexScrubRangeOnly(@Nonnull FDBRecordStore store, @Nonnull AtomicLong recordsScanned) {
+        Index index = common.getIndex();
+        final RecordMetaData metaData = store.getRecordMetaData();
+        final RecordMetaDataProvider recordMetaDataProvider = common.getRecordStoreBuilder().getMetaDataProvider();
+        if (recordMetaDataProvider == null || !metaData.equals(recordMetaDataProvider.getRecordMetaData())) {
+            throw new MetaDataException("Store does not have the same metadata");
+        }
+        final IndexMaintainer maintainer = store.getIndexMaintainer(index);
+        final IndexScrubbingTools<?> tools = maintainer.getIndexScrubbingTools(scrubbingType);
+        if (tools == null) {
+            throw new UnsupportedOperationException("This index does not support scrubbing type " + scrubbingType);
+        }
+
+
+        return indexScrubRangeOnly(store, recordsScanned, index, tools, maintainer.isIdempotent());
+    }
+
+    private <T> CompletableFuture<Boolean> indexScrubRangeOnly(final @Nonnull FDBRecordStore store, final @Nonnull AtomicLong recordsScanned, final Index index, final IndexScrubbingTools<T> tools, boolean isIdempotent) {
+        // scrubbing only readable
+        validateOrThrowEx(store.getIndexState(index) == IndexState.READABLE, "scrubbed index is not readable");
+        // scrubbing only idempotent indexes (at least for now)
+        validateOrThrowEx(isIdempotent, "scrubbed index is not idempotent");
+
+        final IndexingRangeSet rangeSet = getRangeset(store, index);
+        tools.presetCommonParams(index, scrubbingPolicy.allowRepair(), common.getIndexContext().isSynthetic, common.getAllRecordTypes());
```

Comment:
There is a chance that it just seemed like a "nice to have" at the time. Maybe we should deprecate it.

Reply:
Note that the indexer's config API contained 
```
public Builder setRecordTypes(@Nullable Collection<RecordType> recordTypes) 
```
since Fri Dec 14 13:39:45 2018 -0800 (commit ..194208). So it had been there for a while..